src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -91,12 +95,61 @@ public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(Depe
         /// </summary>
         public override async Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(CancellationToken token)
         {
-            ProjectPackages packages = await GetInstalledAndTransitivePackagesAsync(includeTransitiveOrigins: false, token);
-            return packages.InstalledPackages;
+            token.ThrowIfCancellationRequested();
+
+            (PackageSpec packageSpec, string assetsFilePath) = await GetCurrentPackageSpecAndAssetsFilePathSafeAsync(token);
+
+            if (packageSpec == null) // null means project is not nominated
+            {
+                IsInstalledComputationNeeded = true;
+                return Array.Empty<PackageReference>();
+            }
+
+            IReadOnlyList<LockFileTarget> targetsList = null;
+            T installedPackages;
+            if (IsInstalledComputationNeeded)
+            {
+                installedPackages = new T();
+                targetsList = (await GetTargetsListAsync(assetsFilePath, token))?.ToList();
+            }
+            else
+            {
+                if (InstalledPackages == null)
+                {
+                    installedPackages = new T();
+                }
+                else
+                {
+                    // Make a copy of the caches to prevent concurrency issues.
+                    lock (_installedPackagesLock)
+                    {
+                        installedPackages = GetCollectionCopy(InstalledPackages);
+                    }
+                }
+            }
+
+            // get installed packages
+            List<PackageReference> calculatedInstalledPackages = packageSpec
+                .TargetFrameworks
+                .SelectMany(f => ResolvedInstalledPackagesList(f.Dependencies, f.FrameworkName, targetsList, installedPackages))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, FrameworkSorter).First())
+                .ToList();
```

Comment:
Just curious, do we need .ToList(); here?

Reply:
Despite the fact that the API returns IEnumerable, it is preferrable to return a fully evaluated collection. 
If you don't do that and the types change you can get an exception when you use the returned colleciton.

Reply:
To materialize results before sending back.